core/src/main/java/com/radixdlt/consensus/DivergentVertexExecutionDetector.java

```diff
@@ -0,0 +1,157 @@
+ * it may be implemented or used or that it will be appropriate for any specific purpose.
+ *
+ * Neither the licensor nor any current or former employees, officers, directors, partners,
+ * trustees, representatives, agents, advisors, contractors, or volunteers of the Licensor
+ * shall be liable for any direct or indirect, special, incidental, consequential or other
+ * losses of any kind, in tort, contract or otherwise (including but not limited to loss
+ * of revenue, income or profits, or loss of use or data, or loss of reputation, or loss
+ * of any economic or other opportunity of whatsoever nature or howsoever arising), arising
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.consensus;
+
+import com.google.common.hash.HashCode;
+import com.google.common.util.concurrent.RateLimiter;
+import com.radixdlt.consensus.bft.BFTValidatorId;
+import com.radixdlt.consensus.bft.Round;
+import com.radixdlt.monitoring.Metrics;
+import java.util.*;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A utility for detecting divergent vertex execution, i.e. a situation where two (or more)
+ * validators from the validator set have voted for the same vertexId, but their resultant ledger
+ * headers were different.
+ */
+@SuppressWarnings("UnstableApiUsage")
+public final class DivergentVertexExecutionDetector {
```

Comment:
After the refactors, this class now looks like a top-level injectable service:
- (highly recommended) the `Metrics` should be its injected dependency;
- (optional) maybe it could listen on `Vote` and/or `Round` events (instead of relying on those methods being explicitly called from the right places); _I actually dunno if this is feasible because I am still lost around all these events in Java land._

Reply:
I think it's a bit clearer to have the detector being instantiated per consensus instance (which we create manually)

Reply:
but I'll move `metrics` to constructor

Reply:
ahhh the manual instantiation; ok then 👍